bot/exts/info/tags.py

```diff
@@ -24,145 +29,229 @@
-        current, index = 0, 0
-        _search = REGEX_NON_ALPHABET.sub('', search.lower())
-        _targets = iter(REGEX_NON_ALPHABET.split(target.lower()))
-        _target = next(_targets)
-        try:
-            while True:
-                while index < len(_target) and _search[current] == _target[index]:
-                    current += 1
-                    index += 1
-                index, _target = 0, next(_targets)
-        except (StopIteration, IndexError):
-            pass
-        return current / len(_search) * 100
-
-    def _get_suggestions(self, tag_name: str, thresholds: Optional[List[int]] = None) -> List[str]:
-        """Return a list of suggested tags."""
-        scores: Dict[str, int] = {
-            tag_title: Tags._fuzzy_search(tag_name, tag['title'])
-            for tag_title, tag in self._cache.items()
-        }
-
-        thresholds = thresholds or [100, 90, 80, 70, 60]
-
-        for threshold in thresholds:
+                parent_dir = file.relative_to(base_path).parent
+                tag_name = file.stem
+                # Files directly under `base_path` have an empty string as the parent directory name
+                tag_group = parent_dir.name or None
+
+                self.tags[TagIdentifier(tag_group, tag_name)] = Tag(file)
+
+    def _get_suggestions(self, tag_identifier: TagIdentifier) -> list[tuple[TagIdentifier, Tag]]:
+        """Return a list of suggested tags for `tag_identifier`."""
+        for threshold in [100, 90, 80, 70, 60]:
             suggestions = [
-                self._cache[tag_title]
-                for tag_title, matching_score in scores.items()
-                if matching_score >= threshold
+                (identifier, tag)
+                for identifier, tag in self.tags.items()
+                if identifier.get_fuzzy_score(tag_identifier) >= threshold
             ]
             if suggestions:
                 return suggestions
 
         return []
 
-    def _get_tag(self, tag_name: str) -> list:
-        """Get a specific tag."""
-        found = [self._cache.get(tag_name.lower(), None)]
-        if not found[0]:
-            return self._get_suggestions(tag_name)
-        return found
-
-    def _get_tags_via_content(self, check: Callable[[Iterable], bool], keywords: str, user: Member) -> list:
+    def get_fuzzy_matches(self, tag_identifier: TagIdentifier) -> list[tuple[TagIdentifier, Tag]]:
+        """Get tags with identifiers similar to `tag_identifier`."""
+        if tag_identifier.group is None:
+            if len(tag_identifier.name) < 3:
+                return []
+            else:
+                return self._get_suggestions(tag_identifier)
+        else:
+            if len(tag_identifier.group) < 3:
+                suggestions = []
+            else:
+                # Try fuzzy matching with only a name first
+                suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```

Comment:
```suggestion
        if len(tag_identifier.name) < 3 or len(tag_identifier.group) < 3:
            return []
    
        if tag_identifier.group is None:
            return self._get_suggestions(tag_identifier)
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```

Reply:
these do not act the same, the current implementation allows groups shorter than 3 chars to be used to invoke tags as long as the name is longer than 3 and matches

Reply:
Hmm, I see now. What about this instead?

We're removing a level of nesting in the if-statements introducing several dimensions to keep track, by flattening them we get a clear case of when to do what
```suggestion
        if len(tag_identifier.name) < 3:
            return []
        elif tag_identifier.group is None:
            return self._get_suggestions(tag_identifier)

        if len(tag_identifier.group) < 3:
            # No point in fuzzy matching if the group is less than 3 characters
            suggestions = []
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```
Like this we have a pretty clear view..

- If the name is less then 3 characters we return immediately
- If the group is None we can just fuzzy match immediately and return
*I cut them in half here because the first two exit early, the second two continue down the function*
- If the group is less then 3 characters there's no point in fuzzy matching it and just continue with an empty list
- If name is longer then 3 characters, and we have a group over 3 character we fuzzy match the group and continue

Reply:
This won't let us match only with a name if a tag is matched with a long group but a short name, I'm on the edge of that being worth it for the increased code complexity.
On one hand it's relatively unlikely to happen as `!<name> <2chars>` would have to be invoked (maybe an emoji would be an actual case where that can happen) but then the code is also not *that* much less complex without the first if taking care of that case and then the name length check at the end

Reply:
> This won't let us match with only a name if a tag is matched with a long group but a short name

I see now, so we just remove that if-statement and leave it to the code below this. That way we will at least match for groups yeah.
```suggestion
        if tag_identifier.group is None:
            return self._get_suggestions(tag_identifier)

        if len(tag_identifier.group) < 3:
            # No point in fuzzy matching if the group is less than 3 characters
            suggestions = []
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```

Reply:
That would skip the name length check altogether, allowing fuzzying of invocations like `!de` which may just be a typo

Reply:
No it wouldn't, check the code below this comment.

Reply:
The code below won't run when the group is None

Reply:
Ah right, so we can special-case it and just add a second `and len(tag.name) >= 3:`

Reply:
With that `None` groups would go beyond the check for short names

Reply:
I am starting to see why the if-statements look exactly like they do- because I tried to write a suggestion for the code to be exactly the same. That said I came up with something like this:
```py
        if tag_identifier.group is None and len (tag_identifier.name) >= 3:
            return self._get_suggestions(tag_identifier)

        if tag_identifier.group is None or len(tag_identifier.group) < 3:
            # No point in fuzzy matching if the group is less than 3 characters or None
            suggestions = []
        else:
            # Try fuzzy matching with only a name first
            suggestions = self._get_suggestions(TagIdentifier(None, tag_identifier.group))
```
But at this point, this suggestion can be refactored to consider the if-statements below it.

Reply:
```
        suggestions = []

        if tag_identifier.group is not None and len(tag_identifier.group) < 3:
            # Try fuzzy matching with only a name first
            suggestions += self._get_suggestions(TagIdentifier(None, tag_identifier.group))

            if len(tag_identifier.name) >= 3:
                suggestions += self._get_suggestions(tag_identifier)

            return suggestions
```
I can't use the GitHub suggestions for this, because that would duplicate code. But what do you think about this?

Reply:
That should work according to my tests after fixing the inequality in the first if, pushed it in ba2a74ac18f9e40fad917d51ffa2238a340416c0